### PR TITLE
Add support for extended verbosity specification -v"debug +callsite"

### LINKED ISSUE #3768: Intent to refactor: Add call stacks to our debugging output
Have you ever gotten a log message from Cabal and wondered where it came from? Me too! Here's how we're going to fix it:

1. Add a compat shim for GHC.Stack using http://stackoverflow.com/questions/36369252/is-there-a-way-to-syntactically-simulate-empty-constraints-with-ghc-7-8-1
2. Change Verbosity to support extra "trace this specifically." Add a field for annotating debug messages with where they came from
3. Change implementation of debug/info/etc to print the caller filename/lineno if "annotate location of debug messages" is enabled
4. Profit

This is an intent to refactor.

### PR DESCRIPTION
**Pull request by ezyang**
This patch uses CallStack support in GHC 7.10 and GHC 8.0 to make
it possible to get stack traces when we log output.

Example output:

```
ezyang@sabre:~/Dev/cabal-tmp$ /srv/code/cabal-shake/dist-newstyle/build/x86_64-linux/ghc-8.0.1/cabal-install-1.25.0.0/noopt/build/cabal/cabal configure -v"debug +callsite"
Distribution/Simple/Program/Find.hs:77: Searching for ghc in path.
Distribution/Simple/Program/Find.hs:81: Found ghc at /usr/bin/ghc
Distribution/Simple/Program/Builtin.hs:105: ("/usr/bin/ghc",["--numeric-version"])
Distribution/Simple/Program/Builtin.hs:105: /usr/bin/ghc is version 7.10.3
Distribution/Simple/GHC.hs:204: looking for tool ghc-pkg near compiler in /usr/bin
Distribution/Simple/GHC.hs:206: candidate locations:
["/usr/bin/ghc-pkg","/usr/bin/ghc-pkg-ghc-7.10.3","/usr/bin/ghc-pkg-7.10.3","/usr/bin/ghc-pkg"]
Distribution/Simple/GHC.hs:212: found ghc-pkg in /usr/bin/ghc-pkg
Distribution/Simple/Program/Builtin.hs:126: ("/usr/bin/ghc-pkg",["--version"])
Distribution/Simple/Program/Builtin.hs:126: /usr/bin/ghc-pkg is version 7.10.3
Distribution/Simple/Program/Run.hs:163: ("/usr/bin/ghc",["--supported-languages"])
Distribution/Simple/Program/Run.hs:163: ("/usr/bin/ghc",["--info"])
Distribution/Simple/Configure.hs:1257: Reading installed packages...
Distribution/Simple/Program/Run.hs:163: ("/usr/bin/ghc-pkg",["dump","--global","-v0"])
Distribution/Simple/Program/Run.hs:163: ("/usr/bin/ghc-pkg",["dump","--user","-v0"])
Distribution/Simple/Program/Run.hs:163: ("/usr/bin/ghc",["--print-libdir"])
```

This is the bare minimum to make this patch useful: there is
plenty of tuning that can be done.  For example:

* Insertions of withFrozenCallStack can help make the "callsite" output
  more useful, though be careful, we lose all stack information at that point!

* Insertions of 'WithVerbosity', which will let us get deeper stacks
  (at the moment, they are basically always 1-deep.)

Fixes #3768.

CC @23Skidoo

Signed-off-by: Edward Z. Yang <ezyang@cs.stanford.edu>

### CODE REVIEW

**Comment by ezyang** [comment]
Oops, I guess we should use the profiling stack trace, if it is available.

Do you think that `callsite` and `callstack` should stack? Would be easy enough to do.

**Comment by 23Skidoo** [comment]
> Do you think that `callsite` and `callstack` should stack? Would be easy enough to do.

Well, I'd either expect that or the "more detailed" one (`+callstack`) to win.

**Comment by ezyang** [comment]
OK, the default call stacks are now a *lot* more useful. Example:

```
----
CallStack (from HasCallStack):
  findProgramVersion, called at Distribution/Simple/Program/Builtin.hs:126:26 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program.Builtin
  programFindVersion, called at Distribution/Simple/Program/Types.hs:69:8 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program.Types
  configureProgram, called at Distribution/Simple/Program/Db.hs:434:16 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program.Db
  lookupProgramVersion, called at Distribution/Simple/Program/Db.hs:472:3 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program.Db
  requireProgramVersion, called at Distribution/Simple/GHC.hs:118:5 in Cabal-1.25.0.0-inplace:Distribution.Simple.GHC
  configure, called at Distribution/Simple/Configure.hs:1539:14 in Cabal-1.25.0.0-inplace:Distribution.Simple.Configure
  configCompilerEx, called at Distribution/Simple/Configure.hs:1525:27 in Cabal-1.25.0.0-inplace:Distribution.Simple.Configure
  configCompilerAuxEx, called at Main.hs:333:31 in main:Main
/usr/bin/ghc-pkg is version 7.10.3
----
CallStack (from HasCallStack):
  rawSystemStdInOut, called at Distribution/Simple/Program/Run.hs:163:35 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program.Run
  getProgramInvocationOutput, called at Distribution/Simple/Program.hs:161:3 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program
  getProgramOutput, called at Distribution/Simple/GHC/Internal.hs:217:12 in Cabal-1.25.0.0-inplace:Distribution.Simple.GHC.Internal
  getExtensions, called at Distribution/Simple/GHC.hs:143:18 in Cabal-1.25.0.0-inplace:Distribution.Simple.GHC
  configure, called at Distribution/Simple/Configure.hs:1539:14 in Cabal-1.25.0.0-inplace:Distribution.Simple.Configure
  configCompilerEx, called at Distribution/Simple/Configure.hs:1525:27 in Cabal-1.25.0.0-inplace:Distribution.Simple.Configure
  configCompilerAuxEx, called at Main.hs:333:31 in main:Main
("/usr/bin/ghc",["--supported-languages"])
----
CallStack (from HasCallStack):
  rawSystemStdInOut, called at Distribution/Simple/Program/Run.hs:163:35 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program.Run
  getProgramInvocationOutput, called at Distribution/Simple/Program.hs:161:3 in Cabal-1.25.0.0-inplace:Distribution.Simple.Program
  getProgramOutput, called at Distribution/Simple/GHC/Internal.hs:205:13 in Cabal-1.25.0.0-inplace:Distribution.Simple.GHC.Internal
  getGhcInfo, called at Distribution/Simple/GHC.hs:145:14 in Cabal-1.25.0.0-inplace:Distribution.Simple.GHC
  configure, called at Distribution/Simple/Configure.hs:1539:14 in Cabal-1.25.0.0-inplace:Distribution.Simple.Configure
  configCompilerEx, called at Distribution/Simple/Configure.hs:1525:27 in Cabal-1.25.0.0-inplace:Distribution.Simple.Configure
  configCompilerAuxEx, called at Main.hs:333:31 in main:Main
("/usr/bin/ghc",["--info"])
```

File: Cabal/Distribution/Compat/Prelude.hs [hunk 1]
```diff
@@ -122,6 +123,12 @@ import Data.Maybe
 
 import qualified Text.PrettyPrint as Disp
 
+import qualified Prelude as OrigPrelude
+import Distribution.Compat.Stack
+
+type IO a = WithCallStack (OrigPrelude.IO a)
+type NoCallStackIO a = OrigPrelude.IO a
```
Comment (dcoutts) [inline Cabal/Distribution/Compat/Prelude.hs]: This is what I'm sceptical of. Could we reverse the defaults and positively mark things instead?

Looking through this patch it's a little worrying that we may have to think carefully "should this IO helper be marked NoCallstackIO?". Or is there a clear principle here that means it's only a bunch of utility layer things and we don't have to worry about it generally?

**Comment by dcoutts** [comment]
@ezyang I like this. My only concern is whether the IO with callstack vs IO without callstack types are the right way round. When do I have to know to opt-out? Would it be better or worse if it were opt-in?

**Comment by ezyang** [comment]
The principle is that you always want `CallStackIO`, just how in traditional imperative languages, all functions participate in call stack traces. But sometimes GHC throws a redundant constraint error. For this patch, I went through each such occurrence and decided either (1) we should use the CallStack, but don't, or (2) the function never fails (or it's not our function that fails) in which case I annotated it with NoCallStackIO. I would even hazard a guess that some of our NoCallStackIO code could profitably be made with a call stack.

The choice to invert the naming was based on the practicalities of the matter:

1. `ag [^a-zA-Z\.]IO\\W | wc -l` returns 525 (this slightly overreports because it picks up IO from Paths, but not by much). These are not "useless" uses of call stack: in all these cases GHC 8 determined that the constraint was *not* redundant.
2. `NoCallStackIO` returns 80 (this slightly underreports, because FFI signatures also use Prelude IO)

If you are writing a new IO function in the Cabal library, chances are you *want* to use call stack IO. And that's why the defalut is the way it is.

**Comment by dcoutts** [comment]
Ok I think I buy that.

So the cases where the constraint is redundant is where nothing calls any logging function (or die?). And ordinary IO exceptions will of course be annotated with callstacks when possible anyway.

**Comment by ezyang** [comment]
Rebased and merged!

**Pull request closed without merging**